Loader fails to build on MSYS2 (mingw-w64-x86_64) due to compiler-generated memset

In our loader, we have an array initialization that the compiler likes to perform using `memset`: https://github.com/JuliaLang/julia/blob/e949236f29347f3e57a1a3a7d5ec656872ad8347/cli/loader_lib.c#L63

That's not OK, as our loader is compiled without the C standard library. IIUC, that's why we compile with `-ffreestanding`: https://github.com/JuliaLang/julia/blob/e949236f29347f3e57a1a3a7d5ec656872ad8347/cli/Makefile#L10-L11

However, it looks like MSYS2's mingw compiler (GCC 13.2) ignores this:

```
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: ./loader_lib.o: in function `load_library':
C:/Users/Tim/Julia/src/julia/cli/loader_lib.c:63:(.text.startup+0x25): undefined reference to `memset'
collect2.exe: error: ld returned 1 exit status
make: *** [/c/Users/Tim/Julia/src/julia/cli/Makefile:118: /c/Users/Tim/Julia/src/julia/build/usr/bin/libjulia.dll] Error 1
```

I tried a variety of other flags I got from https://github.com/riscv-collab/riscv-gnu-toolchain/issues/758, e.g., `-fno-lto` and `-fno-builtin`, to no avail.

EDIT: looks like `memset` etc are expected to be available, even with `-ffreestanding`:

> GCC requires the freestanding environment provide memcpy, memmove, memset and memcmp.

Workaround:

```c
char path[2*JL_PATH_MAX + 1] /*= {0}*/;
for (int i = 0; i < 2*JL_PATH_MAX + 1; i++)
    path[i] = 0;
```

--- a/cli/loader.h
+++ b/cli/loader.h
@@ -6,24 +6,6 @@
 #include "../src/julia_fasttls.h"
 
 #ifdef _OS_WINDOWS_
-/* We need to reimplement a bunch of standard library stuff on windows,
- * but we want to make sure that it doesn't conflict with the actual implementations
- * once those get linked into this process. */
-#define fwrite loader_fwrite
-#define fputs loader_fputs
-#define exit loader_exit
-#define strlen loader_strlen
-#define wcslen loader_wcslen
-#define strncat loader_strncat
-#define memcpy loader_memcpy
-#define dirname loader_dirname
-#define strchr loader_strchr
-#define malloc loader_malloc
-#define realloc loader_realloc
-#define free loader_free
-#endif
-
-#ifdef _OS_WINDOWS_
 
 #define WIN32_LEAN_AND_MEAN
 #include <windows.h>

--- a/cli/loader_win_utils.c
+++ b/cli/loader_win_utils.c
@@ -12,7 +12,7 @@ static FILE _stderr = { INVALID_HANDLE_VALUE };
 FILE *stdout = &_stdout;
 FILE *stderr = &_stderr;
 
-int __attribute__((weak)) fwrite(const char *str, size_t nchars, FILE *out) {
+int JL_HIDDEN fwrite(const char *str, size_t nchars, FILE *out) {
     DWORD written;
     if (out->isconsole) {
         // Windows consoles do not support UTF-8 (for reading input, though new Windows Terminal does for writing), only UTF-16.
@@ -32,19 +32,19 @@ int __attribute__((weak)) fwrite(const char *str, size_t nchars, FILE *out) {
     return -1;
 }
 
-int __attribute__((weak)) fputs(const char *str, FILE *out) {
+int JL_HIDDEN fputs(const char *str, FILE *out) {
     return fwrite(str, strlen(str), out);
 }
 
-void __attribute__((weak)) *malloc(const size_t size) {
+void JL_HIDDEN *malloc(const size_t size) {
     return HeapAlloc(GetProcessHeap(), HEAP_GENERATE_EXCEPTIONS, size);
 }
 
-void __attribute__((weak)) *realloc(void * mem, const size_t size) {
+void JL_HIDDEN *realloc(void * mem, const size_t size) {
     return HeapReAlloc(GetProcessHeap(), HEAP_GENERATE_EXCEPTIONS, mem, size);
 }
 
-void __attribute__((weak)) free(void* mem) {
+void JL_HIDDEN free(void* mem) {
     HeapFree(GetProcessHeap(), 0, mem);
 }
 
@@ -110,7 +110,7 @@ void setup_stdio() {
     _stderr.isconsole = GetConsoleMode(_stderr.fd, &mode);
 }
 
-void __attribute__((weak)) exit(int code) {
+void JL_HIDDEN exit(int code) {
     ExitProcess(code);
 }
 
@@ -148,21 +148,21 @@ wchar_t *utf8_to_wchar(const char * str) {
     return wstr;
 }
 
-size_t __attribute__((weak)) strlen(const char * x) {
+size_t JL_HIDDEN strlen(const char * x) {
     int idx = 0;
     while (x[idx] != 0)
         idx++;
     return idx;
 }
 
-size_t __attribute__((weak)) wcslen(const wchar_t * x) {
+size_t JL_HIDDEN wcslen(const wchar_t * x) {
     int idx = 0;
     while (x[idx] != 0)
         idx++;
     return idx;
 }
 
-char __attribute__((weak)) *strncat(char * base, const char * tail, size_t maxlen) {
+char JL_HIDDEN *strncat(char * base, const char * tail, size_t maxlen) {
     int base_len = strlen(base);
     int tail_len = strlen(tail);
     for (int idx=base_len; idx<min(maxlen, base_len + tail_len); ++idx) {
@@ -171,21 +171,21 @@ char __attribute__((weak)) *strncat(char * base, const char * tail, size_t maxle
     return base;
 }
 
-void __attribute__((weak)) *memcpy(void * dest, const void * src, size_t len) {
+void JL_HIDDEN *memcpy(void * dest, const void * src, size_t len) {
     for (int idx=0; idx<len; ++idx) {
         ((char *)dest)[idx] = ((const char *)src)[idx];
     }
     return dest;
 }
 
-void __attribute__((weak)) *memset(void *s, int c, size_t n) {
+void JL_HIDDEN *memset(void *s, int c, size_t n) {
   unsigned char* p = s;
   while(n--)
     *p++ = (unsigned char)c;
   return s;
 }
 
-char __attribute__((weak)) *dirname(char * x) {
+char JL_HIDDEN *dirname(char * x) {
     int idx = strlen(x);
     while (idx > 0 && x[idx] != PATHSEPSTRING[0]) {
         idx -= 1;
@@ -205,7 +205,7 @@ char __attribute__((weak)) *dirname(char * x) {
     return x;
 }
 
-char __attribute__((weak)) *strchr(const char * haystack, int needle) {
+char JL_HIDDEN *strchr(const char * haystack, int needle) {
     int idx=0;
     while (haystack[idx] != needle) {
         if (haystack[idx] == 0) {

--- a/cli/loader_win_utils.c
+++ b/cli/loader_win_utils.c
@@ -12,7 +12,7 @@ static FILE _stderr = { INVALID_HANDLE_VALUE };
 FILE *stdout = &_stdout;
 FILE *stderr = &_stderr;
 
-int loader_fwrite(const char *str, size_t nchars, FILE *out) {
+int __attribute__((weak)) fwrite(const char *str, size_t nchars, FILE *out) {
     DWORD written;
     if (out->isconsole) {
         // Windows consoles do not support UTF-8 (for reading input, though new Windows Terminal does for writing), only UTF-16.
@@ -20,10 +20,10 @@ int loader_fwrite(const char *str, size_t nchars, FILE *out) {
         if (!wstr)
             return -1;
         if (WriteConsoleW(out->fd, wstr, wcslen(wstr), &written, NULL)) {
-            loader_free(wstr);
+            free(wstr);
             return written;
         }
-        loader_free(wstr);
+        free(wstr);
     } else {
         // However, we want to print UTF-8 if the output is a file.
         if (WriteFile(out->fd, str, nchars, &written, NULL))
@@ -32,19 +32,19 @@ int loader_fwrite(const char *str, size_t nchars, FILE *out) {
     return -1;
 }
 
-int loader_fputs(const char *str, FILE *out) {
-    return loader_fwrite(str, loader_strlen(str), out);
+int __attribute__((weak)) fputs(const char *str, FILE *out) {
+    return fwrite(str, strlen(str), out);
 }
 
-void * loader_malloc(const size_t size) {
+void __attribute__((weak)) *malloc(const size_t size) {
     return HeapAlloc(GetProcessHeap(), HEAP_GENERATE_EXCEPTIONS, size);
 }
 
-void * loader_realloc(void * mem, const size_t size) {
+void __attribute__((weak)) *realloc(void * mem, const size_t size) {
     return HeapReAlloc(GetProcessHeap(), HEAP_GENERATE_EXCEPTIONS, mem, size);
 }
 
-void loader_free(void* mem) {
+void __attribute__((weak)) free(void* mem) {
     HeapFree(GetProcessHeap(), 0, mem);
 }
 
@@ -110,7 +110,7 @@ void setup_stdio() {
     _stderr.isconsole = GetConsoleMode(_stderr.fd, &mode);
 }
 
-void loader_exit(int code) {
+void __attribute__((weak)) exit(int code) {
     ExitProcess(code);
 }
 
@@ -148,21 +148,21 @@ wchar_t *utf8_to_wchar(const char * str) {
     return wstr;
 }
 
-size_t loader_strlen(const char * x) {
+size_t __attribute__((weak)) strlen(const char * x) {
     int idx = 0;
     while (x[idx] != 0)
         idx++;
     return idx;
 }
 
-size_t loader_wcslen(const wchar_t * x) {
+size_t __attribute__((weak)) wcslen(const wchar_t * x) {
     int idx = 0;
     while (x[idx] != 0)
         idx++;
     return idx;
 }
 
-char * loader_strncat(char * base, const char * tail, size_t maxlen) {
+char __attribute__((weak)) *strncat(char * base, const char * tail, size_t maxlen) {
     int base_len = strlen(base);
     int tail_len = strlen(tail);
     for (int idx=base_len; idx<min(maxlen, base_len + tail_len); ++idx) {
@@ -171,14 +171,21 @@ char * loader_strncat(char * base, const char * tail, size_t maxlen) {
     return base;
 }
 
-void * loader_memcpy(void * dest, const void * src, size_t len) {
+void __attribute__((weak)) *memcpy(void * dest, const void * src, size_t len) {
     for (int idx=0; idx<len; ++idx) {
         ((char *)dest)[idx] = ((const char *)src)[idx];
     }
     return dest;
 }
 
-char * loader_dirname(char * x) {
+void __attribute__((weak)) *memset(void *s, int c, size_t n) {
+  unsigned char* p = s;
+  while(n--)
+    *p++ = (unsigned char)c;
+  return s;
+}
+
+char __attribute__((weak)) *dirname(char * x) {
     int idx = strlen(x);
     while (idx > 0 && x[idx] != PATHSEPSTRING[0]) {
         idx -= 1;
@@ -198,7 +205,7 @@ char * loader_dirname(char * x) {
     return x;
 }
 
-char * loader_strchr(const char * haystack, int needle) {
+char __attribute__((weak)) *strchr(const char * haystack, int needle) {
     int idx=0;
     while (haystack[idx] != needle) {
         if (haystack[idx] == 0) {

